### PR TITLE
Add files watcher for rebuild preview

### PR DESCRIPTION
**Pull request by phuongnq**
### Ticket reference or full description of what's in the PR
https://github.com/craftercms/craftercms/issues/6223

### CODE REVIEW

File: src/main/java/org/craftercms/engine/service/context/SiteContextFactory.java [hunk 1]
```diff
@@ -451,6 +454,25 @@ public SiteContext createContext(String siteName) {
         }
     }
 
+    /**
+     * Resolve root folder path
+     * @param siteName site name
+     * @return the root folder absolute path
+     */
+    public String resolvedRootFolderPath(String siteName) throws URISyntaxException {
```
Comment (jmendeza) [inline src/main/java/org/craftercms/engine/service/context/SiteContextFactory.java:462]: Why `resolved` (vs `resolve`) ?
`resolvedRootFolderPath`
`resolveRootFolderPath`

**Review by jmendeza — CHANGES_REQUESTED** [review]

File: src/main/java/org/craftercms/engine/service/context/SiteContextFactory.java (hunk 1 above)
Comment (phuongnq) [reply src/main/java/org/craftercms/engine/service/context/SiteContextFactory.java]: Updated to use `resolve`. Thank you!

**Review by jmendeza — APPROVED** [review]

**Review by sumerjabri — CHANGES_REQUESTED** [review]
Questions:

1. What happens if someone pulls a large changeset with files with different dates? Will it rebuild many times even though they're in close temporal proximity (same time)?
2. Are we creating one thread per rebuild? How is thread management done?

My suggested algorithm is:
Main idea: Trigger once for a large set of changes, but never wait more than 1 second
Ingredients:
- A counter with a configurable limit (default 5)
- A configurable timeout (default 200 milliseconds)

- When the watcher triggers, set a counter to 1 and sleep for 200 milliseconds
- Wake up and check:
  - Is the counter 5 (one second has passed)? Then trigger a rebuild
  - Has anything else changed (more changes since we slept)? If so, increment the counter and sleep for 200 milliseconds. If nothing has changed, then trigger a rebuild
- All of the above happens in the same thread, or a thread pool with limits is used

File: pom.xml [hunk 2]
```diff
@@ -376,6 +376,12 @@
             </exclusions>
         </dependency>
 
+        <dependency>
+            <groupId>io.methvin</groupId>
+            <artifactId>directory-watcher</artifactId>
+            <version>${directoryWatcher.version}</version>
```
Comment (sumerjabri) [inline pom.xml:382]: directory-watcher.version

Comment (phuongnq) [reply pom.xml]: Updated. Thank you!

File: src/main/java/org/craftercms/engine/service/context/SiteContextManager.java [hunk 3]
```diff
@@ -197,6 +210,79 @@ public void createContexts(boolean concurrent) {
...
+                            case MODIFY: {
+                                String hashValue = event.hash() != null ? event.hash().asString() : "none";
+                                logger.info("File watcher event kind: '{}'. File affected: '{}'. Hash value: '{}'", event.eventType(), event.path(), hashValue);
+                                // Only process if the hash is different from the last processed hash value,
+                                // or the event hash is null in the case of DELETE
+                                // This means the modified time has been updated from last processed event or the file is deleted
+                                // This prevents multiple context rebuild for batch files change with same modified date such as from a git pull
+                                String lastProcessedHash = directoryWatcherLastProcessedHash.get(siteName);
+                                if (lastProcessedHash == null || event.hash() == null || !lastProcessedHash.equals(hashValue)) {
+                                    Lock lock = siteLockFactory.getLock(siteName);
```
[the remark this answers is not in the document]
Comment (phuongnq) [reply src/main/java/org/craftercms/engine/service/context/SiteContextManager.java]: Updated. Thank you

**Comment by phuongnq** [comment]
@sumerjabri I have updated the apply the algorithm.

> 1. What happens if someone pulls a large changeset with files with different dates? Will it rebuild many times even though they're in close temporal proximity (same time)?

=> As far as from my test. If the pull has CREATE, MODIFY actions, the modified date are the same so we only have one event. But for DELETE, we have multiple events since the hash is null in the case of DELETE. Your suggested algorithm helps with this case (multiple DELETE to trigger only one rebuild)

> 2. Are we creating one thread per rebuild? How is thread management done?

=> With updating the algorithm. I have separated threads to the watch event and the rebuild task. 
* Watch events are based on modified date and this will only update a counter of how many events have been occurred from last rebuild.
* Rebuild tasks are implemented as of the suggested algorithm based on how many events have been occurred from the watch event thread. The rebuild if any will be perform within the same thread of rebuild checking task (one thread for one site) with the lock/unlock method.

File: src/main/java/org/craftercms/engine/service/context/SiteContextManager.java [hunk 4]
```diff
@@ -197,6 +255,139 @@ public void createContexts(boolean concurrent) {
...
+                                    } finally {
+                                        siteLock.unlock();
+                                    }
+                                } else {
+                                    logger.info("File watch for hash '{}' has already processed. No action required.", hashValue);
+                                }
+                                break;
+                            }
+                            default:
+                                logger.info("File watcher unhandled event kind: '{}'. File affected: '{}'.", event.eventType(), event.path());
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/engine/service/context/SiteContextManager.java:306]: Switch to debug. Also, "File watcher unhandled event type" (not "kind").

File: src/main/java/org/craftercms/engine/service/context/SiteContextManager.java [hunk 5]
```diff
@@ -197,6 +255,139 @@ public void createContexts(boolean concurrent) {
...
+                                        if (counter == null) {
+                                            counter = new AtomicInteger(0);
+                                        }
+                                        counter.getAndIncrement();
+                                        directoryWatcherCounter.put(siteName, counter);
+                                    } finally {
+                                        siteLock.unlock();
+                                    }
+                                } else {
+                                    logger.info("File watch for hash '{}' has already processed. No action required.", hashValue);
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/engine/service/context/SiteContextManager.java:301]: Switch to debug.

File: src/main/java/org/craftercms/engine/service/context/SiteContextManager.java [hunk 6]
```diff
@@ -197,6 +255,139 @@ public void createContexts(boolean concurrent) {
...
+                    DirectoryWatcher oldWatcher = directoryWatcherRegistry.remove(siteName);
+                    oldWatcher.close();
+                } finally {
+                    siteLock.unlock();
+                }
+            }
+            directoryWatcherRegistry.put(siteName, watcher);
+            watcher.watchAsync();
+        } catch (Exception e) {
+            logger.error("Error while creating watcher for site: " + siteName, e);
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/engine/service/context/SiteContextManager.java:325]: Switch to a String.format instead of string concatenation.

**Review by sumerjabri — CHANGES_REQUESTED** [review]

File: src/main/java/org/craftercms/engine/service/context/SiteContextManager.java (hunk 4 above)
Comment (phuongnq) [reply src/main/java/org/craftercms/engine/service/context/SiteContextManager.java]: Updated. Thank you!

File: src/main/java/org/craftercms/engine/service/context/SiteContextManager.java (hunk 5 above)
Comment (phuongnq) [reply src/main/java/org/craftercms/engine/service/context/SiteContextManager.java]: Updated. Thank you!

File: src/main/java/org/craftercms/engine/service/context/SiteContextManager.java (hunk 6 above)
Comment (phuongnq) [reply src/main/java/org/craftercms/engine/service/context/SiteContextManager.java]: Updated. Thank you!

**Review by sumerjabri — APPROVED** [review]

**Pull request merged**